tests/util/Component.js

```diff
@@ -0,0 +1,85 @@
+'use strict';
+
+const assert = require('assert');
+const eslint = require('eslint');
+const Components = require('../../lib/util/Components');
+
+const ruleTester = new eslint.RuleTester({
+  parserOptions: {
+    ecmaVersion: 2018,
+    sourceType: 'module',
+    ecmaFeatures: {
+      jsx: true,
+    },
+  },
+});
+
+describe('Components', () => {
+  describe('static detect', () => {
+    function makeTest(name, test, done) {
+      return it(name, () => {
+        const rule = Components.detect((context, components, util) => ({
+          'Program:exit'(node) {
+            done(name, context, components, util, node);
+          },
+        }));
+
+        const tests = {
+          valid: [Object.assign(test, {
+            settings: {
+              react: {
+                version: 'detect',
+              },
+            },
+          })],
+          invalid: [],
+        };
+        ruleTester.run(name, rule, tests);
+      });
+    }
+
+    makeTest('Stateless Function Components', {
+      code: `import React from 'react'
+        function MyStatelessComponent() {
+          return <></>
+        }`,
+    }, (name, _context, components) => {
+      assert(components.length() === 1, `${name} should be detected component`);
+      Object.values(components.list()).forEach((component) => {
+        assert(
+          component.node.id.name === 'MyStatelessComponent',
+          `${name} should be named MyStatelessComponent`
+        );
+      });
+    });
+
+    makeTest('Class Components', {
+      code: `import React from 'react'
+        class MyClassComponent extends React.Component {
+          render() {
+            return <></>
+          }
+        }`,
+    }, (name, _context, components) => {
+      assert(components.length() === 1, `${name} should be detected component`);
+      Object.values(components.list()).forEach((component) => {
+        assert(
+          component.node.id.name === 'MyClassComponent',
+          `${name} should be named MyClassComponent`
+        );
+      });
+    });
+
+    makeTest('React imports', {
+      code: 'import React, { useCallback, useState } from \'react\'',
+    }, (_name, _context, components) => {
+      const defaultReactImport = components.getDefaultReactImport();
+      assert(defaultReactImport.local.name === 'React', 'default React import identifier should be "React"');
+
+      const namedReactImports = components.getNamedReactImports();
+      assert(namedReactImports.length === 2, 'named React import identifiers should be "useCallback" and "useState"');
+      assert(namedReactImports[0].local.name === 'useCallback', 'named React import identifiers should be "useCallback" and "useState"');
+      assert(namedReactImports[1].local.name === 'useState', 'named React import identifiers should be "useCallback" and "useState"');
```

Comment:
```suggestion
      assert.deepEqual(
        namedReactImports.length,
        [
          { local: { name: 'useCallback' } },
          { local: { name: 'useState' } },
        ], 
        'named React import identifiers should be "useCallback" and "useState"'
      );
```

Reply:
This doesn't really work in this instance because the returned identifier nodes have so much other stuff attached to them, that we end up with a failing `deepEqual` and an unreadable error message.

I could assert over all that stuff, but it's not really the point of the test.

```
      named React import identifiers should be "useCallback" and "useState"
Occurred while linting <input>:1
Rule: "import React, { useCallback, useState } from 'react'"
      + expected - actual

       [
         {
      -    "imported": {
      -      "loc": {
      -        "end": {
      -          "column": 27
      -          "line": 1
      -        }
      -        "start": {
      -          "column": 16
      -          "line": 1
      -        }
      -      }
      -      "name": "useCallback"
      -      "parent": [Circular]
```

Reply:
Maybe deepEqual to `namedReactImports.map(x => x.local.name)`?